platform/mbed_error.h

```diff
@@ -66,12 +845,220 @@
  *
  *
  */
+    
+MBED_DEPRECATED_SINCE("mbed-os-5.9", "This function has been deprecated, please use one of SET_WARNING/SET_ERROR macros or one of set_warning/set_error functions" )
 
-#ifdef __cplusplus
-extern "C" {
-#endif
 void error(const char* format, ...);
```

Comment:
Wow! Deprecating `error`! I think this is the wrong place for this.

How about we keep the new implementation but wait to deprecate until the logging changes? That way `debug` and `error` can be deprecated at the same time, and users will get less deprecation churn.

And even then we **really** shouldn't deprecate `error` or `debug` until we're sure logging is stable. There's no real rush to deprecate.

Reply:
I agree with @geky here, we should not deprecate `error`. Current `error` API does both logging and handling of error, it is more for critical errors which will result in application termination.
However with error handling we would not terminate on all errors.

Reply:
SET_ERROR does terminate the application, just as how error did. SET_WARNING does not terminate the application but will log the error context.

Reply:
> By the way, error logging is not same as logging or tracing implementation we are doing,

Error logging (not creation/codes) should come under the same log feature be it mbed_trace or unified logging.  APIs for dumping data on UART should not be part of error handling.

Reply:
We should not deprecate `error` until we deprecate `debug`

Reply:
Unfortunately, we cannot use mbed_trace as it uses too much resources+stdlibs, and the idea is to have an implementation to output to serial/uart without using much of system resources. If you have noticed the mbed_error_print is minimal implementation which was already there as part of exception handling, but we moved it under mbed_error_report to be used for error handling as well. By the error handling uses crit sect, so it will not corrupt other output.

Reply:
Note that nothing should be hard-coded to serial/uart - we've got full console abstraction now. There should be no `serial_putc`s in any of these paths, and one of the outstanding jobs is to remove them all in favour of `write(STDERR_FILENO)`.

Reply:
@kjbracey-arm - I see your point, but the intention here is to use as much less resources as possible
as this is the same path for exception handling. So, using those high level abstracted calls may not work in all cases for example when we are dealing with an fatal exception. Just to give you more info, we are planning/thinking on a recovery mode feature where in we reset the system without re-initing the piece memory having the error info and then on reboot we can print that info out through write calls normally. At that we can get rid special handling when dealing with exception handling/errors handling.

Reply:
I'm not saying we need to adopt mbed-trace right now, I'm saying we should not deprecate `error` until we deprecate `debug`. You have an implementation or `error` here, so just remove the deprecation.

There's no urgency to deprecate `error` and users don't like making multiple changes.